[BOT-260] sem search consistency, gating and indexer loops

wotbrew opened the pull request:
## Problem

See BOT-260 for even more context, but to summarize. We are trying to solve the following problem:

> Ensure that the index eventually gets populated with all the searchable documents on an instance, even if errors occur.

- If indexing fails due to embedding unavailability we drop those writes
- Ingest can provide documents out of order, meaning you never converge on a state where indexed documents reflect the instances searchable documents

There is a gradient of solutions to these sorts of problems with varying trade-offs. This PR provides a solution but perhaps it leans a little 'more sophisticated' than other short term options. We should carefully consider whether this solution and its implications are suitable for release now. I do lean toward this, only because its easier to do this early before we launch - and it in my view changes the solution space for so many other problems.

## Description 

Online index maintainance hooks now issue updates through a document 'gating' mechanism. 
A single asyncronous indexer process per cluster then updates/deletes documents in the pgvector indexes by querying the gating table.

The API level indexing methods work as before, so expect no major changes to tests or the implementation of other namespaces, new `gate-` functions are added to pgvector-api for interaction with this new mechanism.

### Mechanism

The gate table produces a more predictable order for document changes (ultimately derived from the t2 updated_at). Writes from any number of nodes/threads can hit the gate table concurrently and the logical order is respected such that older documents do not replace new ones.

Consumers can then be notified of document changes (new/updated/deleted) efficiently. Downstream indexing can proceed independently, its processing is no longer coupled to the ingest queue.

This decoupling has some important consequences:
- Indexing can take as long as it likes (if say, embedder is unavailable) without causing unbounded memory usage in ingest (unbounded delay queue)
- Limits likelyhood of delays to the ingest producer limiting the availability or quality of the other indexes (e.g. appdb)
- Less chance of data loss if process is lost during indexing
- Reindexing from source does not imply redundantly recomputing everything, we only need to compute embeddings / index where our (instance) state has diverged from the expected (gate) state

Objective: The ordering provides a _higher likelyhood_ of the index converging on a consistent state with infrequent reindexes. 

**TLDR; Indexing can fail and we will resume where we left off**

## Additional discussion

This PR only focuses on the semantic search side of the equation. We can still lose data and so on - it only should not happen as often. If semsearch is later hosted as an external service we should be pretty robust taking writes from many machines over HTTP whose ordering would be even less likely to reflect any logical order than is currently the case.

Assume any kondos / accidental whitespace diffs will be addressed pre merge.

### TODO on this PR (or immediately after merge)

- No tests of the job definition / quartz level (though the implementation is reasonably tested)
- No tests of the very important timeout on gate write

### Further work

- Redo the idle policy (note in conversations), I know what I want but I think I'd rather do a follow up and this will not hurt.
- Looking at the `reindex`/`init` calls with gating in mind
- Improve observability
- Determine correct policy for non transient errors (e.g. malformed documents or bugs in our code). Should we flag those docs as 'no good' and proceed to try and index what we can? I suspect a more partially complete index is still preferable to a less partially complete one.
- Better model switching coordination, right now you might have to wait for a redundant indexer run before the new model is picked up
- GC policy for very old deletes to avoid unbounded growth if the id space can grow large.
- Need to look at racey start ups (related to other problems we are working on) and sem search initialization. Its a broader problem, and right now I think its ok to crash or use the wrong active index until its fixed within 10 secs - 1 hour by the rescheduling of the quartz job.

wotbrew committed:
Add document gating

Adds a document gating mechanism and async indexing job.

- Allows indexing to recover if it encounters a transient error.
- Avoids data loss during indexing due to process death
- Provides an ordering which is more closely related to logical record
validity times rather than write time into the index.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/gate.clj:
```diff
@@ -0,0 +1,90 @@
+                             :from   [(keyword (:gate-table-name index-metadata))]
+                             ;; the earliest timestamp where we might expect to find new documents
+                             :where  [:gated_at :>= [:min
+                                                     ;; lag our last-poll slightly, as we may not have seen everything
```
After sleep this makes less sense, cant we just have the last poll - lag given its one clock (drop the `min`).

Only thing the :last-seen expr does in this `min` is make it so lag configuration is irrelevant when it is < the confidence window.

Makes stuff more confusing IMO with little beneft

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/index_metadata.clj:
```diff
@@ -11,29 +11,39 @@
     [:index_version :int :not-null]
-    [:index_created_at :timestamp-with-time-zone :not-null]]
+    [:index_created_at :timestamp-with-time-zone :not-null]
+    [:indexer_last_poll :timestamp-with-time-zone :not-null]
```
watermark state

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -0,0 +1,236 @@
+          (log/info "Starting new indexing thread")
+          (let [new-indexing-state  (init-indexing-state metadata-row)
+                ;; virtual thread could also be good here, most time will be blocking on I/O
+                new-indexing-thread (Thread. ^Runnable (fn [] (indexing-loop pgvector index-metadata index new-indexing-state)))]
```
todo name, daemon

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -0,0 +1,236 @@
+  (volatile! {:shutting-down false
+              :processes     {}}))
+
+(defn shutdown!
```
shutdown hook

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer_heartbeat.clj:
```diff
@@ -0,0 +1,52 @@
+  SemanticSearchIndexer
+  [_]
+  @shutdown-hook
+  (semantic-search.indexer/heartbeat! proc-map (semantic.env/get-pgvector-datasource!) (semantic.env/get-index-metadata)))
```
This assumes node affinity which quartz does not provide - will need to figure that out or do this a different way (could dedupe/order immediately through the gate and have a fixup job instead)

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -0,0 +1,164 @@
+(defn on-indexing-idle [indexing-state]
+  (let [{:keys [last-poll-count last-indexed-count]} @indexing-state
+        novelty-ratio (if (zero? last-poll-count) 0 (/ last-indexed-count last-poll-count))]
+    (cond
```
this survived too long from sketch stage but its too cute I'll probably tight-loop until buffer is full, poll cost ought to be marginal compared to embed/index and we can revisit if `:limit` ends up being huge for some reason

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer.clj:
```diff
@@ -0,0 +1,66 @@
+(def ^:private ^Duration startup-delay (Duration/parse "PT10S"))
+(def ^:private ^Duration run-frequency (Duration/parse "PT10S"))
+
+;; TODO skip/do not run if no ee flag - how does this
```
need to not run job if we are not semantic search enabled or if we have no db configuration

wotbrew replied:
I think what I'll do is check the flag in the job execution, but only schedule the job if the pg vector url is available (since we expect that to not ever change at runtime).

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer.clj:
```diff
@@ -0,0 +1,66 @@
+        (.interrupt execution-thread)))))
+
+(def ^:private ^Duration startup-delay (Duration/parse "PT10S"))
+(def ^:private ^Duration run-frequency (Duration/parse "PT10S"))
```
This only represents a indexing gap when the gate indexer is cold. Individual runs will take up to an hour given the current configuration - indexing is eventually yielded to quartz - so jobs do not run indefinitely. 

Also note that the cold early exit is defined in the indexer loop, so that if we do not see anything new in some time, we also yield back to quartz. 

I did not want to have an infinite quartz job, intuitively it felt that would cause havoc. Feel free to call me out on this - maybe that its fine!

FWIW Normally with this kind of thing you would do a bit of leader election for each index or something like that: ultimately the goal is ensuring multiple nodes do not clobber the same index.

appleby replied:
Is there any benefit to having the startup-delay be randomized at all? I feel like I've seen slack threads in the past were cloud restarts a bunch of instances at the same time and they all kickoff some startup job at ~ the same time and smash the appdb (or pgvector db in this case). I guess reindexing on startup takes long enough that a small amount of randomization is unlikely to help much.

appleby replied:
> I did not want to have an infinite quartz job

1+ Something we could consider for the future, but I think having it yield back is a good place to start

wotbrew replied:
> is there any benefit to having the startup-delay be randomized at all

I am not sure this is needed in this case, there should only be one instance of the job running at a time?

wotbrew replied:
No no it is a good point! I am not sure how the virtualization of pgvector is going to work, its magic to me. I probably need to learn about that before I can say anything useful!

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/pgvector_api.clj:
```diff
@@ -79,11 +81,50 @@
   (let [{:keys [index]} (ensure-active-index-state pgvector index-metadata)]
     (semantic.index/upsert-index! pgvector index documents)))
 
+(defn gate-updates!
```
It'd prefer a single `gate!` function rather han seperate update/delete functions (simple proxy, maybe only adding h the partitioning behaviour).  This would be both a richer and simpler calling pattern IMO regarding the interleaved update/delete gate! actually supports - but it would involve hoisting up the gate doc mapping. Maybe in the future.

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/gate_test.clj:
```diff
@@ -0,0 +1,297 @@
+          (is (= (map comparable-gate-row previous-state)
+                 (map comparable-gate-row (get-gate-rows pgvector index-metadata))))))
+
+      (testing "documents are not deleted if timestamp <= current"
```
this testing doc is totally wrong 😅

wotbrew replied:
e2ab9a0e0332

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/gate_test.clj:
```diff
@@ -0,0 +1,297 @@
+      (testing "documents are not deleted if already deleted, regardless of timestamp"
+        (is (= 0 (sut pgvector index-metadata [(delete d1 t2)]))))
+
+      (testing "documents are not undeleted if timestamp < "
```
ditto doc

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj:
```diff
@@ -0,0 +1,454 @@
+             (fn []
+               (loop []
+                 (when-not (.isInterrupted (Thread/currentThread))
+                   (try
```
Ah, I remember it was needed to test "exceptions are bubbled out of loop, will not loop forever" 
when some other stuff was more complicated. Delete!

wotbrew replied:
2599b0586dd0

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj:
```diff
@@ -0,0 +1,454 @@
+                              :thread
+                              (doto (Thread.
+                                     (fn []
+                                       (loop []
```
ditto I think this loop is redundant, clearly no `recur`. Double check below asserts hold regardless and if not why the hell did they not fail!

wotbrew replied:
2599b0586dd0

wotbrew committed:
Fix some test docstrings

wotbrew committed:
Fix some loop tests

wotbrew committed:
Schedule indexer only if the instance has the :semantic-search feature

wotbrew committed:
Avoid redundant replay of the very last gate entry when the quartz job fires

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj:
```diff
@@ -0,0 +1,463 @@
+              (is (= (ex-message ex) (ex-message @caught-ex)))
+              (vreset! step-ex nil)
+              (wait-for-calls)
+              (testing "sanity: check loop no longer throwing"
```
redundant, delete (this survived a previous pass when I simulated the quartz job here a bit more). 

We should `(is)` check the returns of `(wait-for-calls)` otherwise it just times out and the error is not reported. Go through and fix these cases where the return is ignored.

wotbrew replied:
107c573e4159

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj:
```diff
@@ -0,0 +1,463 @@
+   (semantic.index/create-index-table-if-not-exists! pgvector index)
+   (fn [_] (semantic.index/drop-index-table! pgvector index))))
+
+(defn- spy [f]
```
I think there is a `semantic.tu/spy` now - reuse that

noahmoss commented on enterprise/backend/src/metabase_enterprise/semantic_search/pgvector_api.clj:
```diff
@@ -79,11 +81,50 @@
+  [pgvector index-metadata documents]
+  (let [now (Instant/now)]
+    (transduce
+     (partition-all (min 512 semantic.gate/max-batch-size))
```
related to my comment above, if `max-batch-size` is a defsetting you could enforce a minimum in a custom getter, if that's what you're intending here. although do we need a hard-coded lower bound on this?

wotbrew replied:
oh man that should be `(max)`!

wotbrew replied:
👍  on the setting, good idea!

I am not quite sure what you are getting at with the `min` / lower-bound comment?

The `(min)` takes the lower value, so it puts a high bound on what it can be (the max-batch-size). So I think this code makes sense? Perhaps you are talking about something else?

The reason for the seperate variables is because the desired batch size and the maximum enforced by the gate are set their values for different reasons. They just happen to have the same value right now because they are both complete guesses until measurement!

- the desired batch size is most likely weighted toward throughput during reindexes
- max is basically a proxy for 'can be done before the gate write times out')

**TLDR;** Yes on the (bounded) setting for batch size - I will add a follow up for this? I do not think it blocks?

> **aside** The actual maximum is hard to define - it is likely a function of postgres load and configuration - batch size, document sizes, record locality,  its always going to be a bit of a guess!
What I think would be cool is attempt a write at the desired batch size, if we timeout - half the batch size and retry.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -0,0 +1,196 @@
+            update-candidates)))
+
+(defn update-candidate-filter [candidate-filter update-candidates]
+  (let [{:keys [capacity tree]} candidate-filter
```
Will add some extra docs for clarity!

wotbrew replied:
I realise after sleep that the whole candidate filter thing does not require any near this much complexity. You can simply remember the set of keys you have seen last time and dedupe against that. Will delete!

wotbrew replied:
fd7a2e7f5d1b way easier!

wotbrew commented on enterprise/backend/test/metabase_enterprise/semantic_search/test_util.clj:
```diff
@@ -89,15 +89,17 @@
+     :metadata-table-name   (format fmt "metadata")
+     :control-table-name    (format fmt "control")
+     :gate-table-name       (format fmt "gate")
      :index-table-qualifier fmt}))
```
Unless you are referring to that, I think the qualifier being a format string was intentional IIRC so that you could chose either/both a suffix/prefix, like `test_%s_3242343`.

My 2pence on where I might personally look to go here because I do think it is a bit yuk:

With @lbrdnk  looking at $liquibase rather than qualifying the tables we could look at creating schemas or something like that for isolation. With metadata I fell into qualifying the names like this due to the prior index table name variables, but it was not though through enough. It is nice in tests to be able to qualify like this, would let us run tests in parallel if we want without multiple pgs - but schema would also work just fine for isolation.

I would still keep table names variables for the actual index tables ( I think here it still makes sense), I would not for metadata roots unless we want to do some kind multitenancy thing longer term where being super flexible (indirect!) about how to address the data might bring some advantage (in which case which physical database, which schema, which tables might all benefit from being variables).

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/gate.clj:
```diff
@@ -0,0 +1,261 @@
+  ;; i.e (the WHERE must be re-evaluated against the UPDATED row if another transaction commits concurrently).
+  (jdbc/with-transaction [tx pgvector {:isolation :read-committed}]
+    (when-some [gate-document-batch (->> gate-document-batch
+                                         (sort-by :updated_at)
```
Good spot, makes sense! I think in this case java.time.Instant comes from the defaults for updated_at if the source doc does not have one.

 I'll grep the appdb schema for updated_at and ensure we can handle all the JDBC / java.time type zoo that emerges - and that our default updated at is also compatible.
 
 It also makes me think of a test idea, I should load all the searchable docs for the test appdb and gate them, all the tests right now use synthetic data for the tests which means it never runs in to anything like this.

wotbrew committed:
Add a quick candidate-filter byte equality test

wotbrew committed:
Refine loop interrupt/error test (get rid of old gunk)

wotbrew committed:
Use shared `tu/spy` instead of copy-pasta

wotbrew committed:
Replace candidate filter with simple set of last seen

Use TEXT for hash for eq convenience.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -0,0 +1,196 @@
+  (let [{:keys [watermark candidate-filter]} @indexing-state
+
+        move-to-next-watermark
+        (fn [poll-result]
```
Great catch 👍

wotbrew replied:
@bronsa What do you think of 4ce9e9735877? Agree this is better?

bronsa replied:
yeah lgtm

wotbrew committed:
Avoid future foot guns with a (letfn) refactor of indexing-step

Avoids issues that might later be introduced by the 'accidental' closure of (actual) variables that might later be changed during this side-effecting operation. In particular the move-to-next-watermark local function.

^:once was suggested but a hoisted letfn that can only close over args (which keeps the mutable indexing-state at the forefront of ones mind) keeps us honest.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer.clj:
```diff
@@ -0,0 +1,66 @@
+(def ^:private ^Duration run-frequency (Duration/parse "PT10S"))
+
+(defmethod task/init! ::SemanticSearchIndexer [_]
+  (when (premium-features/has-feature? :semantic-search)
```
I mean, I definitely fiddle with the db at runtime at the REPL but I can always manually reschedule the job - more for users expectations.

wotbrew replied:
06327835b862 as a stop gap

wotbrew committed:
Check if engine enabled and supported during indexer task

Previously was only checking if the db was configured.

wotbrew committed:
Be more careful about Instant and timestamp representation.

Preferred is sql.Timestamp for now due to the default setup of postgres / JDBC.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/gate.clj:
```diff
@@ -0,0 +1,261 @@
+  (let [last-poll-time    (or (:last-poll watermark) Instant/EPOCH)
+        last-seen-time    (or (:gated_at (:last-seen watermark)) Instant/EPOCH)
+        confidence-time   (.minus (Instant/ofEpochMilli (inst-ms last-poll-time)) lag-tolerance)
+        gate-min          (if (< (inst-ms confidence-time) (inst-ms last-seen-time))
```
1c69e6940d49

wotbrew committed:
Fix up some asserts that still reflected byte hashes

noahmoss reviewed (approved):
Did another high-level pass, the core components make sense to me and I don't think any [issues](https://metaboat.slack.com/archives/C07SJT1P0ET/p1754586888496499?thread_ts=1754503040.252499&cid=C07SJT1P0ET) I found while testing are blocking, so stamping to unblock (assuming tests/lint errors will be fixed before this goes in).

@appleby / @bronsa should probably give a 👍 too though since it's such a big PR

bronsa reviewed (approved):
happy with the changes, good to merge once CI is green

wotbrew committed:
Add a quick e2e index everything test.

Find horrific pagination bug. Fix that.

We did not account for having enough ties that our buffer was full. This was likely in tests even with the high precision of the pg clock because:
- we coerced stuff to Instant which lost precision
- we generally did not account for tie breaks in pagination (bad form)

So we are paginating properly using the secondary id as a tie break. Within the confidence time only, as otherwise this would be dangerous (it is right to stall if you see too much until the lag time has elapsed). This stops stalls if lots of documents share a buffer.

Will follow up with another test that forces time collisions to assert this tie break behaviour independently.

appleby reviewed (approved):
Thanks for the detailed writeup in the PR description 👍 

Hooked it up to stats db and seemed to be working well for me locally :shipit:

wotbrew committed:
Test fixes

wotbrew committed:
Joy

wotbrew committed:
Add indexer docstring
